### DISCUSSION TITLE
how to disable asm language for openssl in ics-openvpn android build

### DISCUSSION THREAD

**Discussion by FxamarAboali** [schwabe/ics-openvpn#1676] [General]
hi 
l edit aes class c .
when i want to build app i can't see my modify because openssl use asm asmply language by default .
how to disable no-asm language for openssl build

**Comment by schwabe** [comment]
I am not really comfortible answering this question. Sure that is possible but at the same time I feel someone how messes with something like AES or other OpenSSL internals should be already be able to do that because otherwise messing with these things is really a good thing to do

**Reply by FxamarAboali** [reply]
ok I know that, but I modify class with my cypher, I can't add my cypher class in OpenSSL.

I disabled asm language in OpenSSL by use make .\config --no-asm when I build OpenSSL server, and everything is ok.
but I can't disable asm in ics-openvpn android I don't know how to do (make .\config --no-asm) in android. 
please if you know tell me.

Thank you.

**Reply by schwabe** [reply]
Add own cipher under its own name instead of modifying an existing one. Your answer raises more questions than its answer. Disabling assembler in my app for OpenSSL is certainly possible and to emulate that OpenSSL options but for me to spend time how to do that and explain it, I need to be convinced that what you are trying to do is not awful and that it is a good idea to help you.

**Comment by FxamarAboali** [comment]
I added an encryption for my name in OpenSSl, but openvpn did not recognize it, so I modified an existing encryption in order to complete my university graduation project because the project will be delivered very soon and I cannot delay it any longer. 
So I had to take this step because it was urgent, and later I will try to create my own encryption. 
I hope you will help me complete my graduation project. 
just I whant to disable asm in app .

I thank you for your cooperation, my dear brother.

**Reply by schwabe** [reply]
openssl is built using the openssl.cmake. You need to add OPENSSL_NO_ASM to the defines. After that you probably need to fix a number of other issues that come up :)

**Reply by FxamarAboali** [reply]
I added set(OPENSSL NO ASM ON) under enable_language(ASM) in openssl.cmake but not affected anything, can you explain more details about that.
 my knowledge is limited about this.
thank you.

**Reply by schwabe** [reply]
I am afraid that to pull this off under Android with the current CMake based build system, you will need a lot more knowledge than that. The step you are currently struggling with is just the first of many changes that will be needed to make to get no asm building under Android working.

The build system for Android and OpenSSL has been basically written from scratch and it just does not support building with no-asm right, so there is no simple toggle or switch that you can set to enable/disable ASM. Adding building without ASM is something, someone needs to implement in this build system and that requires knowledge of cmake and C/C++ based build systems.

I would suggest speaking with your advisor and pointing out that this part might be too big of a scope for your project and look into alternatives.

**Reply by FxamarAboali** [reply]
ok how can add my class cypher in OpenSSL and ics-openvpn to use

**Reply by schwabe** [reply]
OpenVPN will use the EVP_CIPHER_fetch OpenSSL method to fetch a cipher configured in the configuration. SO you need to ensure that you modify OpenSSL to know about the new name cipher of yours.

**Reply by FxamarAboali** [reply]
I can't do anything for the last Soulutions.

**Comment by FxamarAboali** [comment]
how to compiler c language to asm languge for openssl build in ics-openvpn

**Reply by schwabe** [reply]
I think explaining how compilers and toolchains work is really outside the scope of this project. That is espeically not something that is specific to this project.

**Reply by FxamarAboali** [reply]
how to solve my problem,  please help me

**Reply by schwabe** [reply]
I think I given enough pointers now how to achieve what you are trying to do. But I cannot teach you how cmake, C or OpenSSL work. Or do your unversity assignment for you. That would fall into the realm of professional training and/or commissioned work. Please understand that I don't provide those things for free.

**Reply by FxamarAboali** [reply]
Ok brother.
what can i do for you? To give me that?

Would you like to exchange information, make a payment, or donate to charity? And how much money.

**Reply by schwabe** [reply]
You have not even explained why you are trying to modify AES or what your overall goal is or how that modified client would be used.

**Reply by FxamarAboali** [reply]
I have my own encryption algorithm. I want to add it to OpenSSL to use in OpenVPN and connect to it for all encryption. Just that goal.

**Reply by schwabe** [reply]
People do not normally add their own cipher (not cypher!) algorithm to OpenSSL. Normally unless you have very good reasons, this is a fool's errand and I don't want to waste my time on some misguided attempt at cryptography. So unless you give me a bigger picture and why this is a good idea in your and what your true goal is, you will not getting any help from me.

**Reply by FxamarAboali** [reply]
I want you to build ics-openvpn with no-asm only.
The goal is to try to create my final project at my university using my algorithm,

 If you can do that for me and give you the deal that we both agree on fine. Or I change your ics-openvpn app with anther app, you are the loser, as you want.

I am waiting for your approval?

**Reply by schwabe** [reply]
Then I prefer to be the looser. I do not want to be part of this thing you are doing. Paying someone else to do part of their theses is at least unethical in some places and outright cheating/forbidden in other places.

**Reply by FxamarAboali** [reply]
I'm not talking to change your app no. I mean I will use another app. 
This is my meaning.

**Reply by FxamarAboali** [reply]
i apologize if my spelling not correctly . because my language not english.

ok

**Reply by schwabe** [reply]
I prefer someone to use another app for cryptographically questionable undertakings like modifying AES and not adding a new cipher into OpenSSL than to have my name/app associated with what I feel is damaging my reputation anyway.

**Comment by FxamarAboali** [comment]
The external answer you were given is that it change custom cmake files.
how to try to do a ./Configure no-asm ans ./Çonfigure

**Reply by schwabe** [reply]
OpenVPN does not use the normal OpenSSL build system. Instead it uses `openssl.cmake`. You need change that openssl.cmake.
